gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaWrapper.java

```diff
@@ -0,0 +1,342 @@
+      for (TopicMetadata topicMetadata : topicMetadataList) {
+        List<KafkaPartition> partitions = getPartitionsForTopic(topicMetadata);
+        filteredTopics.add(new KafkaTopic(topicMetadata.topic(), partitions));
+      }
+      return filteredTopics;
+    }
+
+    private List<KafkaPartition> getPartitionsForTopic(TopicMetadata topicMetadata) {
+      List<KafkaPartition> partitions = Lists.newArrayList();
+
+      for (PartitionMetadata partitionMetadata : topicMetadata.partitionsMetadata()) {
+        partitions.add(new KafkaPartition.Builder().withId(partitionMetadata.partitionId())
+            .withTopicName(topicMetadata.topic()).withLeaderId(partitionMetadata.leader().id())
+            .withLeaderHost(partitionMetadata.leader().host()).withLeaderPort(partitionMetadata.leader().port())
+            .build());
+      }
+      return partitions;
+    }
+
+    private List<TopicMetadata> getFilteredMetadataList(Set<String> blacklist, Set<String> whitelist) {
+      List<TopicMetadata> filteredTopicMetadataList = Lists.newArrayList();
+
+      //Try all brokers one by one, until successfully retrieved topic metadata (topicMetadataList is non-null)
+      for (String broker : KafkaWrapper.this.brokers) {
+        filteredTopicMetadataList = fetchTopicMetadataFromBroker(broker, blacklist, whitelist);
+        if (filteredTopicMetadataList != null) {
+          break;
+        }
+      }
+
+      if (filteredTopicMetadataList == null) {
+        throw new RuntimeException(
+            "Fetching topic metadata from all brokers failed. See log warning for more information.");
+      }
+      return filteredTopicMetadataList;
+    }
+
+    private List<TopicMetadata> fetchTopicMetadataFromBroker(String broker, Set<String> blacklist, Set<String> whitelist) {
+
+      List<TopicMetadata> topicMetadataList = fetchTopicMetadataFromBroker(broker);
+      List<TopicMetadata> filteredTopicMetadataList = Lists.newArrayList();
+      for (TopicMetadata topicMetadata : topicMetadataList) {
+        if (MessageBasedSource.survived(topicMetadata.topic(), blacklist, whitelist)) {
+          filteredTopicMetadataList.add(topicMetadata);
+        }
+      }
+      return filteredTopicMetadataList;
+    }
+
+    private List<TopicMetadata> fetchTopicMetadataFromBroker(String broker, String... selectedTopics) {
+      LOG.info(String.format("Fetching topic metadata from broker %s", broker));
+      SimpleConsumer consumer = null;
+      try {
+        consumer = getSimpleConsumer(broker);
+        for (int i = 0; i < NUM_TRIES_FETCH_TOPIC; i++) {
+          try {
+            return consumer.send(new TopicMetadataRequest(Arrays.asList(selectedTopics))).topicsMetadata();
+          } catch (Exception e) {
+            LOG.warn(String.format("Fetching topic metadata from broker %s has failed %d times. Reason: %s", broker,
+                i + 1, e.getMessage()));
+            try {
+              Thread.sleep((long) (Math.random() * (i + 1) * 1000));
```

Comment:
Good suggestion. This part is inherited from Camus, but I think your method is better.